src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/24.diff.sql

```diff
@@ -0,0 +1,146 @@
+                                                            INNER JOIN sys.partition_functions AS pf WITH (NOLOCK) ON pf.function_id = prv.function_id
+                                                        WHERE pf.name = N'PartitionFunction_ResourceChangeData_Timestamp'
+                                                            AND SQL_VARIANT_PROPERTY(prv.Value, 'BaseType') = 'datetime2'
+                                                            AND CAST(prv.value AS datetime2(7)) < DATEADD(HOUR, DATEDIFF(HOUR, 0, @lastProcessedUtcDateTime), 0)
+                                                        ORDER BY prv.boundary_id DESC);    
+
+    /* It ensures that it will not check resource changes in future partitions. */
+    DECLARE @endDateTimeToFilter AS datetime2(7) = DATEADD(HOUR, 1, SYSUTCDATETIME());
+
+    WITH PartitionBoundaries
+    AS (
+        SELECT CAST(prv.value as datetime2(7)) AS PartitionBoundary FROM sys.partition_range_values AS prv WITH (NOLOCK)
+            INNER JOIN sys.partition_functions AS pf WITH (NOLOCK) ON pf.function_id = prv.function_id
+        WHERE pf.name = N'PartitionFunction_ResourceChangeData_Timestamp'
+                AND SQL_VARIANT_PROPERTY(prv.Value, 'BaseType') = 'datetime2'
+                AND (
+                    /* Normal logic when prior watermark exists, Grab prior partition to ensure we do not miss data that was written across a partition boundary,
+                       and includes partitions until current one to ensure we keep moving to the next partition when some partitions do not have any resource changes. */
+                    CAST(prv.value AS datetime2(7)) BETWEEN @precedingPartitionBoundary AND @endDateTimeToFilter
+                    OR 
+                    /* It happens when no prior watermark exists or the last processed datetime of prior watermark is older than the last retention datetime.
+                       Uses the partition anchor datetime as the last processed DateTime. */
+                    @precedingPartitionBoundary IS NULL AND CAST(prv.value AS datetime2(7)) BETWEEN CONVERT(datetime2(7), N'1970-01-01T00:00:00.0000000') AND @endDateTimeToFilter
+                )
+    )
+    SELECT TOP(@pageSize) Id,
+        Timestamp,
+        ResourceId,
+        ResourceTypeId,
+        ResourceVersion,
+        ResourceChangeTypeId
+    FROM PartitionBoundaries AS p
+    CROSS APPLY (
+        /* Given the fact that Read Committed Snapshot isolation level is enabled on the FHIR database,
+           using TABLOCK and HOLDLOCK table hints to avoid skipping resource changes 
+           due to interleaved transactions on the resource change data table. */
+        SELECT TOP(@pageSize) Id,
+            Timestamp,
+            ResourceId,
+            ResourceTypeId,
+            ResourceVersion,
+            ResourceChangeTypeId
+        /* Acquires and holds the table lock to prevent new resource changes from being created during the select query execution. */
```

Comment:
Suggestion: Extend comment to indicate that without the TABLOCK & HOLDLOCK key and row locks will be used which can result in missed reads if an insert occurs in range that was previously read and the locks were released on the read.

Reply:
Added an additional comment.